README.md
# Extension fluiddisplay

This extension is a Tesseract component. It is a Data Consumer based on Fluid templates.
Refer to the [manual for more details](https://docs.typo3.org/typo3cms/extensions/fluiddisplay/).


ChangeLog
2017-06-06 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Released to TER as version 2.1.0

2017-06-03 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Added some debug information

2017-05-26 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Verified compatibility with TYPO3 CMS 8

2016-04-24 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Added composer manifest, resolves #75891
	* Reformatted code according to PSR-2, resolves #75892
	* Cleaned up code, resolves #75893
	* Moved language files to XLIFF, resolves #75894
	* Released to TER as version 2.0.0, resolves #58679

2015-09-27 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Verified compatibility with TYPO3 CMS 7, resolves #69939
	* Added reference to description field, resolves #70133
	* Moved documentation to reST, resolves #69940

2014-05-09 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Verified compatibility with TYPO3 CMS 6.2, resolves #58678
	* Released to TER as version 1.4.0, resolves #58679

2013-06-21 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Released to TER as version 1.3.0

2013-05-23 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Added a substitute page title view helper, resolves #46234
	* Cleaned up the sample template, resolves #48488
	* Verified compatibility with TYPO3 6.0 and 6.1, resolves #47519

2013-05-22 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Added FAL support, resolves #48471

2012-06-29 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Added definition of search fields, resolves #38483

2012-06-26 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Cleaned up some code, verified compatibility with TYPO3 4.7, resolves #38391

2012-01-08 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Replaced reference to parent object with new, general member variable, resolves #33053

2010-09-27 Fabien Udriot (Ecodev)  <fabien.udriot@ecodev.ch>

	* first release

ext_tables.php
<?php
if (!defined ('TYPO3_MODE')) {
	die ('Access denied.');
}

\TYPO3\CMS\Core\Utility\ExtensionManagementUtility::allowTableOnStandardPages('tx_fluiddisplay_displays');

// Add context sensitive help (csh) for this table
\TYPO3\CMS\Core\Utility\ExtensionManagementUtility::addLLrefForTCAdescr(
	'tx_fluiddisplay_displays',
	'EXT:' . $_EXTKEY . '/Resources/Private/Language/locallang_csh_txfluiddisplaydisplays.xlf'
);

// Register sprite icon for fluiddisplay table
/** @var \TYPO3\CMS\Core\Imaging\IconRegistry $iconRegistry */
$iconRegistry = \TYPO3\CMS\Core\Utility\GeneralUtility::makeInstance(\TYPO3\CMS\Core\Imaging\IconRegistry::class);
$iconRegistry->registerIcon(
        'tx_fluiddisplay-display',
        \TYPO3\CMS\Core\Imaging\IconProvider\BitmapIconProvider::class,
        [
            'source' => 'EXT:fluiddisplay/Resources/Public/Icons/FluidDisplay.png'
        ]
);


ext_localconf.php
<?php
if (!defined('TYPO3_MODE')) {
    die ('Access denied.');
}

// Register as Data Consumer service
// Note that the subtype corresponds to the name of the database table
\TYPO3\CMS\Core\Utility\ExtensionManagementUtility::addService(
        'fluiddisplay',
        // Service type
        'dataconsumer',
        // Service key
        'tx_fluiddisplay_dataconsumer',
        array(
                'title' => 'Fluid-based Data Consumer',
                'description' => 'Data Consumer for recordset-type data structures, based on Fluid templating',

                'subtype' => 'tx_fluiddisplay_displays',

                'available' => true,
                'priority' => 50,
                'quality' => 50,

                'os' => '',
                'exec' => '',

                'className' => \Tesseract\Fluiddisplay\Component\DataConsumer::class,
        )
);



Classes/Component/DataConsumer.php
<?php
namespace Tesseract\Fluiddisplay\Component;

/*
 * This file is part of the TYPO3 CMS project.
 *
 * It is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License, either version 2
 * of the License, or any later version.
 *
 * For the full copyright and license information, please read the
 * LICENSE.txt file that was distributed with this source code.
 *
 * The TYPO3 project - inspiring people to share!
 */

use Tesseract\Tesseract\Service\FrontendConsumerBase;
use Tesseract\Tesseract\Tesseract;
use Tesseract\Tesseract\Utility\Utilities;
use TYPO3\CMS\Core\Messaging\FlashMessage;
use TYPO3\CMS\Core\Utility\GeneralUtility;
use TYPO3\CMS\Fluid\View\StandaloneView;

/**
 * This class is the actual data consumer for extension fluiddisplay.
 * It performs a rendering of the data structure using a Fluid template.
 *
 * @author Francois Suter (Cobweb) <typo3@cobweb.ch>
 * @author Fabien Udriot <fabien.udriot@ecodev.ch>
 * @package TYPO3
 * @subpackage tx_fluiddisplay
 */
class DataConsumer extends FrontendConsumerBase
{

    public $tsKey = 'tx_fluiddisplay';
    public $extKey = 'fluiddisplay';
    protected $conf;
    protected $table; // Name of the table where the details about the data display are stored
    protected $uid; // Primary key of the record to fetch for the details
    protected $structure = array(); // Input standardised data structure
    protected $result = ''; // The result of the processing by the Data Consumer
    protected $counter = array();

    /**
     * Resets values for a number of properties.
     *
     * This is necessary because services are managed as singletons.
     *
     * @return void
     */
    public function reset()
    {
        $this->structure = array();
        $this->result = '';
        $this->uid = '';
        $this->table = '';
        $this->conf = array();
    }

    /**
     * Returns the filter data.
     *
     * @return array
     */
    public function getFilter()
    {
        return $this->filter;
    }

    // Data Consumer interface methods

    /**
     * Returns the type of data structure that the Data Consumer can use.
     *
     * @return string Type of used data structures
     */
    public function getAcceptedDataStructure()
    {
        return Tesseract::RECORDSET_STRUCTURE_TYPE;
    }

    /**
     * Indicates whether the Data Consumer can use the type of data structure requested or not.
     *
     * @param string $type Type of data structure
     * @return boolean True if it can use the requested type, false otherwise
     */
    public function acceptsDataStructure($type)
    {
        return $type === Tesseract::RECORDSET_STRUCTURE_TYPE;
    }

    /**
     * Passes a data structure to the Data Consumer.
     *
     * @param array $structure Standardised data structure
     * @return void
     */
    public function setDataStructure($structure)
    {
        $this->structure[$structure['name']] = $structure;
    }

    /**
     * Gets the data structure.
     *
     * @return array $structure Standardised data structure
     */
    public function getDataStructure()
    {
        return $this->structure;
    }

    /**
     * Returns the result of the rendering.
     *
     * @return mixed The result of the Data Consumer's work
     */
    public function getResult()
    {
        return $this->result;
    }

    /**
     * Sets the result. Useful for hooks.
     *
     * @param mixed $result Predefined result
     * @return void
     */
    public function setResult($result)
    {
        $this->result = $result;
    }

    /**
     * Starts the rendering using the Fluid engine.
     *
     * @return void
     */
    public function startProcess()
    {

        // Get the full path to the template file
        try {
            $filePath = Utilities::getTemplateFilePath($this->consumerData['template']);

            // Instantiate a Fluid stand-alone view and load the template file
            $view = GeneralUtility::makeInstance(StandaloneView::class);
            $view->setTemplatePathAndFilename($filePath);
            // Assign the Tesseract Data Structure
            $view->assign('datastructure', $this->structure);
            // Define a hook allowing pre-processing of the view before rendering
            if (is_array($GLOBALS['TYPO3_CONF_VARS']['EXTCONF'][$this->extKey]['preProcessView'])) {
                foreach ($GLOBALS['TYPO3_CONF_VARS']['EXTCONF'][$this->extKey]['preProcessView'] as $className) {
                    $preProcessor = GeneralUtility::getUserObj($className);
                    $preProcessor->preProcessView($view, $this);
                }
            }
            // Send some information to debug output
            $this->controller->addMessage(
                    $this->extKey,
                    'Template file: ' . $filePath,
                    '',
                    FlashMessage::INFO
            );
            $this->controller->addMessage(
                    $this->extKey,
                    'Received data structure',
                    '',
                    FlashMessage::INFO,
                    $this->structure
            );
            // Render the result
            $this->result = $view->render();

        } catch (\Exception $e) {
            $this->controller->addMessage(
                    $this->extKey,
                    $e->getMessage() . ' (' . $e->getCode() . ')',
                    'Error processing the view',
                    FlashMessage::ERROR
            );
        }
    }
}


Configuration/TCA/Overrides/tt_content.php
<?php
if (!defined ('TYPO3_MODE')) {
	die ('Access denied.');
}

// Add a wizard for adding a fluiddisplay
$addTemplateDisplayWizard = array(
	'type' => 'script',
	'title' => 'LLL:EXT:fluiddisplay/Resources/Private/Language/locallang_db.xlf:wizards.add_fluiddisplay',
	'script' => 'wizard_add.php',
	'module' => array(
		'name' => 'wizard_add'
	),
	'icon' => 'EXT:fluiddisplay/Resources/Public/Icons/AddFluidDisplayWizard.png',
	'params' => array(
		'table' => 'tx_fluiddisplay_displays',
		'pid' => '###CURRENT_PID###',
		'setValue' => 'set'
	)
);
$GLOBALS['TCA']['tt_content']['columns']['tx_displaycontroller_consumer']['config']['wizards']['add_fluiddisplay'] = $addTemplateDisplayWizard;

// Register fluiddisplay with the Display Controller as a Data Consumer
$GLOBALS['TCA']['tt_content']['columns']['tx_displaycontroller_consumer']['config']['allowed'] .= ',tx_fluiddisplay_displays';


Configuration/TCA/tx_fluiddisplay_displays.php
<?php
if (!defined ('TYPO3_MODE')) {
	die ('Access denied.');
}

return array(
	'ctrl' => array(
		'title' => 'LLL:EXT:fluiddisplay/Resources/Private/Language/locallang_db.xlf:tx_fluiddisplay_displays',
		'label' => 'title',
		'descriptionColumn' => 'description',
		'tstamp' => 'tstamp',
		'crdate' => 'crdate',
		'cruser_id' => 'cruser_id',
		'default_sortby' => 'ORDER BY title',
		'delete' => 'deleted',
		'enablecolumns' => array(
			'disabled' => 'hidden',
		),
		'searchFields' => 'title,description,template',
		'typeicon_classes' => array(
			'default' => 'tx_fluiddisplay-display'
		),
	),
	'interface' => array(
		'showRecordFieldList' => 'hidden,title,description'
	),
	'columns' => array(
		'hidden' => array(
			'exclude' => 1,
			'label' => 'LLL:EXT:lang/locallang_general.xlf:LGL.hidden',
			'config' => array(
				'type' => 'check',
				'default' => '0'
			)
		),
		'title' => array(
			'exclude' => 0,
			'label' => 'LLL:EXT:fluiddisplay/Resources/Private/Language/locallang_db.xlf:tx_fluiddisplay_displays.title',
			'config' => array(
				'type' => 'input',
				'size' => '30',
				'eval' => 'required,trim',
			)
		),
		'description' => array(
			'exclude' => 0,
			'label' => 'LLL:EXT:fluiddisplay/Resources/Private/Language/locallang_db.xlf:tx_fluiddisplay_displays.description',
			'config' => array(
				'type' => 'text',
				'cols' => '40',
				'rows' => '4',
			)
		),
		'template' => array(
			'exclude' => 0,
			'label' => 'LLL:EXT:fluiddisplay/Resources/Private/Language/locallang_db.xlf:tx_fluiddisplay_displays.template',
			'config' => array(
				'type' => 'input',
				'size' => '30',
				'eval' => 'required,trim',
				// TODO: remove old wizard definition and replace with new one when TYPO3 v7 compatibility is dropped
				'wizards' => array(
					'_PADDING' => 2,
					'link' => array(
						'type' => 'popup',
						'title' => 'Link',
						'icon' => 'link_popup.gif',
						'module' => array(
							'name' => 'wizard_element_browser',
							'urlParameters' => array(
								'mode' => 'wizard'
							)
						),
						'JSopenParams' => 'height=600,width=700,status=0,menubar=0,scrollbars=1',
						'params' => array(
							'blindLinkOptions' => 'page,url,mail,spec,folder',
							'allowedExtensions' => $GLOBALS['TYPO3_CONF_VARS']['SYS']['textfile_ext'],
						),
					)
				)
			)
		),
	),
	'types' => array(
		'0' => array('showitem' => 'hidden, title;;1, template')
	),
	'palettes' => array(
		'1' => array('showitem' => 'description')
	)
);


Classes/ViewHelpers/SubstitutePageTitleViewHelper.php
<?php
namespace Tesseract\Fluiddisplay\ViewHelpers;

/*
 * This file is part of the TYPO3 CMS project.
 *
 * It is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License, either version 2
 * of the License, or any later version.
 *
 * For the full copyright and license information, please read the
 * LICENSE.txt file that was distributed with this source code.
 *
 * The TYPO3 project - inspiring people to share!
 */

use TYPO3\CMS\Fluid\Core\ViewHelper\AbstractViewHelper;

/**
 * This class is a Fluiddisplay view helper for the Fluid templating engine.
 *
 * It takes the content of the view helper tag and sets it as the page title,
 * returning it or not, depending on the settings.
 *
 * = Examples =
 *
 * Assuming that the namespace has been declared as follows:
 * {namespace fluiddisplay = Tesseract\Fluiddisplay\ViewHelpers}
 *
 * <code title="Simple string with expression to evaluate">
 * <fluiddisplay:substitutePageTitle output="true">{record.header}</fluiddisplay:substitutePageTitle>
 * </code>
 * <output>
 * Foo (assuming the "header" property of object "record" was "Foo")
 * </output>
 * NOTE: if the output property is not set or set to false, the view helper will return an empty string.
 *
 * @package TYPO3
 * @subpackage tx_fluiddisplay
 * @author Francois Suter <typo3@cobweb.ch>
 */
class SubstitutePageTitleViewHelper extends AbstractViewHelper
{

    /**
     * Substitutes the page title with the view helper's content
     *
     * @param boolean $output Set to true if the content of the tag should be returned
     * @return string The content of the tag or an empty string
     */
    public function render($output = false)
    {
        $output = (boolean)$output;
        // Render the content and set it as the page title
        $content = $this->renderChildren();
        if (!empty($content)) {
            $GLOBALS['TSFE']->page['title'] = $content;
            $GLOBALS['TSFE']->indexedDocTitle = $content;
        }
        // If output is requested, return content. Otherwise return empty string.
        return ($output) ? $content : '';
    }

}


ext_emconf.php
<?php

/***************************************************************
 * Extension Manager/Repository config file for ext "fluiddisplay".
 *
 * Auto generated 06-06-2017 17:34
 *
 * Manual updates:
 * Only the data in the array - everything else is removed by next
 * writing. "version" and "dependencies" must not be touched!
 ***************************************************************/

$EM_CONF[$_EXTKEY] = array (
  'title' => 'Fluid-based Data Consumer - Tesseract project',
  'description' => 'Use Fluid-based templates to display any kind of data returned by a Data Provider. More info on http://www.typo3-tesseract.com.',
  'category' => 'fe',
  'author' => 'Francois Suter (Cobweb) / Fabien Udriot',
  'author_email' => 'typo3@cobweb.ch',
  'state' => 'stable',
  'uploadfolder' => 0,
  'createDirs' => '',
  'clearCacheOnLoad' => 0,
  'author_company' => '',
  'version' => '2.1.0',
  'constraints' => 
  array (
    'depends' => 
    array (
      'typo3' => '7.6.0-8.99.99',
      'tesseract' => '2.0.0-0.0.0',
    ),
    'conflicts' => 
    array (
    ),
    'suggests' => 
    array (
    ),
  ),
  '_md5_values_when_last_written' => 'a:38:{s:9:"ChangeLog";s:4:"dfa6";s:11:"LICENSE.txt";s:4:"6404";s:9:"README.md";s:4:"5962";s:13:"composer.json";s:4:"2480";s:12:"ext_icon.png";s:4:"1ce9";s:17:"ext_localconf.php";s:4:"f183";s:14:"ext_tables.php";s:4:"a10e";s:14:"ext_tables.sql";s:4:"9c07";s:34:"Classes/Component/DataConsumer.php";s:4:"1062";s:53:"Classes/ViewHelpers/SubstitutePageTitleViewHelper.php";s:4:"78d0";s:46:"Configuration/TCA/tx_fluiddisplay_displays.php";s:4:"e1a6";s:42:"Configuration/TCA/Overrides/tt_content.php";s:4:"064f";s:56:"Configuration/TCA/Overrides/tx_fluiddisplay_displays.php";s:4:"1842";s:26:"Documentation/Includes.txt";s:4:"6d5f";s:23:"Documentation/Index.rst";s:4:"2d2d";s:26:"Documentation/Settings.yml";s:4:"5ece";s:33:"Documentation/Developer/Index.rst";s:4:"42c0";s:44:"Documentation/Images/FluidDisplayElement.png";s:4:"c2c4";s:59:"Documentation/Images/FluidDisplayElementV8BrokenPreview.png";s:4:"4ae3";s:56:"Documentation/Images/FluidDisplayElementV8InputField.png";s:4:"1f88";s:49:"Documentation/Images/SampleTemplateController.png";s:4:"7286";s:44:"Documentation/Images/SampleTemplateQuery.png";s:4:"6d8b";s:36:"Documentation/Installation/Index.rst";s:4:"5b84";s:36:"Documentation/Introduction/Index.rst";s:4:"5584";s:37:"Documentation/KnownProblems/Index.rst";s:4:"66af";s:28:"Documentation/User/Index.rst";s:4:"5131";s:42:"Documentation/User/FluidPartials/Index.rst";s:4:"a240";s:40:"Documentation/User/InputScreen/Index.rst";s:4:"efb4";s:45:"Documentation/User/LocalizedStrings/Index.rst";s:4:"05b4";s:43:"Documentation/User/SampleTemplate/Index.rst";s:4:"a54a";s:40:"Documentation/User/ViewHelpers/Index.rst";s:4:"d6c1";s:67:"Resources/Private/Language/locallang_csh_txfluiddisplaydisplays.xlf";s:4:"5d5c";s:43:"Resources/Private/Language/locallang_db.xlf";s:4:"c9fb";s:48:"Resources/Public/Icons/AddFluidDisplayWizard.png";s:4:"9cf0";s:39:"Resources/Public/Icons/FluidDisplay.png";s:4:"cd04";s:21:"Samples/locallang.xlf";s:4:"d907";s:27:"Samples/sampleTemplate.html";s:4:"62fc";s:35:"Samples/Partials/samplePartial.html";s:4:"3bef";}',
);

